xcc: fix warning about unsupported compiler/linker flags on no standard defines/libraryes

dcpleung opened the pull request:
XCC does not support the GCC flags to not include standard include files and libraries (e.g. -undef, -nodefaultlibs). So make these into parameters and XCC compiler configuration can override them.

tyrantlizard77 reviewed (approved):
my cmake-fu is weak but this looks reasonable

nashif commented:
@mped-oticon looks like the things you wanted to do in general for compiler/option abstraction

mped-oticon commented:
> @mped-oticon looks like the things you wanted to do in general for compiler/option abstraction

This is definitely a step in the right direction, but what I will do will be higher level - to move as much logic out of `CmakeLists.txt`  as possible.

SebastianBoe commented on cmake/compiler/xcc.cmake:
```diff
@@ -0,0 +1,7 @@
+  -nostdlib
+)
+
+set(NOSYSDEF_CFLAG "")
```
Isnt' this dangerous?

Then on xcc it will be 

-D__SOMETHING_THAT_SHOULD_HAVE_BEEN_UNDEFINED

SebastianBoe replied:
My mistake, I thought -undef was a different flag ...

SebastianBoe commented:
zephyr_ld_options (as opposed to zephyr_link_libararies) is supposed to test the flag and then only include it if it is supported by the toolchain.

So if everything is working as it should, there should be no warning and no need for this patch. I don't have xcc, could you post the warning you are getting or reproduce the warning with the SDK?

SebastianBoe commented on CMakeLists.txt:
```diff
@@ -165,9 +165,7 @@ endif()
 zephyr_ld_options(
-  -nostartfiles
-  -nodefaultlibs
-  -nostdlib
```
Since we are touching this, might as well fix it at the same time.

"-nostdlib" is equivalent to "-nostartfiles -nodefaultlibs"

so specifying all three is redundant, we should remove -nostartfiles and -nodefaultlibs.

From the docs:

> -nostartfiles
> Do not use the standard system startup files when linking. The standard system libraries are used normally, unless -nostdlib, -nolibc, or -nodefaultlibs is used.
> 
> -nodefaultlibs
> Do not use the standard system libraries when linking. Only the libraries you specify are passed to the linker, and options specifying linkage of the system libraries, such as -static-libgcc or -shared-libgcc, are ignored. The standard startup files are used normally, unless -nostartfiles is used.
> 
> The compiler may generate calls to memcmp, memset, memcpy and memmove. These entries are usually resolved by entries in libc. These entry points should be supplied through some other mechanism when this option is specified.
> 
> -nostdlib
> Do not use the standard system startup files or libraries when linking. No startup files and only the libraries you specify are passed to the linker, and options specifying linkage of the system libraries, such as -static-libgcc or -shared-libgcc, are ignored.
> 
> The compiler may generate calls to memcmp, memset, memcpy and memmove. These entries are usually resolved by entries in libc. These entry points should be supplied through some other mechanism when this option is specified.
> 
> One of the standard libraries bypassed by -nostdlib and -nodefaultlibs is libgcc.a, a library of internal subroutines which GCC uses to overcome shortcomings of particular machines, or special needs for some languages. (See Interfacing to GCC Output in GNU Compiler Collection (GCC) Internals, for more discussion of libgcc.a.) In most cases, you need libgcc.a even when you want to avoid other standard libraries. In other words, when you specify -nostdlib or -nodefaultlibs you should usually specify -lgcc as well. This ensures that you have no unresolved references to internal GCC library subroutines. (An example of such an internal subroutine is __main, used to ensure C++ constructors are called; see collect2 in GNU Compiler Collection (GCC) Internals.)

SebastianBoe replied:
Coincidentally, when those two flags are dropped we no longer need the NOSYSDEF_CFLAG since xcc supports -nostdlib.

dcpleung replied:
I have uploaded a new revision with those two removed.

nashif commented:
installing xcc is documented here: https://docs.zephyrproject.org/latest/boards/xtensa/intel_s1000_crb/doc/intel_s1000.html

You can download it and get a license directly.

SebastianBoe commented on cmake/dts.cmake:
```diff
@@ -92,7 +92,7 @@ if(CONFIG_HAS_DTS)
     ${DTC_INCLUDE_FLAG_FOR_DTS}  # include the DTS source and overlays
     -I${ZEPHYR_BASE}/dts/common
-    -undef -D__DTS__
+    ${NOSYSDEF_CFLAG} -D__DTS__
```
While you are patching this, could you move -D__DTS__ to the next line?

Right now it reads like "Don't define __DTS__".

dcpleung replied:
Changed in new revision.

SebastianBoe commented on cmake/toolchain/xcc.cmake:
```diff
@@ -1,6 +1,6 @@
 set(TOOLCHAIN_HOME /opt/xtensa/XtDevTools/install/tools/$ENV{TOOLCHAIN_VER}/XtensaTools)
 
-set(COMPILER gcc)
+set(COMPILER xcc)
```
Doing this would be inconsistent with how other toolchains solve the same problem.

We should instead do as xcc did with 'CC' for instance.

Have gcc.cmake do

if(NOT DEFINED NOSYSDEF_CFLAG)
set(NOSYSDEF_CFLAG -undef)
endif()

and then do

set(NOSYSDEF_CFLAG "")

in toolchain/xcc.cmake

dcpleung replied:
Modified the patches to do this instead.

dcpleung committed:
cmake: parameterized flags to not include standard defines/files

Not all compiles/linkers support the GCC flags to not include standard
defines, include files and libraries. So make these parameters such
that the toolchain can define them when supported.

Also, according to documentation, -nostdlib does the same for both
-nostartfiles and -nodefaultlibs. So remove the redundant ones.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
cmake: xcc: fix compile warnings and errors

XCC does not support the "-undef" flag so set NOSYSDEF_CFLAGS to
empty string to fix compile warning.

Also, XCC does not supply the macro __SIZEOF_LONG__ which breaks
the build. So define it.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung commented:
XCC does emit warnings:

```
xt-xcc WARNING:  -nostartfiles is ignored
xt-xcc WARNING:  -nodefaultlibs is ignored
```

It seems like cmake does not pick up these warnings.
